pkg/controllers/updaterun/execution.go

```diff
@@ -0,0 +1,482 @@
+			klog.ErrorS(failedErr, "The cluster has failed to be updated", "clusterStagedUpdateRun", updateRunRef)
+			return 0, fmt.Errorf("%w: %s", errStagedUpdatedAborted, failedErr.Error())
+		}
+		if condition.IsConditionStatusTrue(clusterUpdateSucceededCond, updateRun.Generation) {
+			// The cluster has been updated successfully.
+			finishedClusterCount++
+			continue
+		}
+		// The cluster is either updating or not started yet.
+		binding := toBeUpdatedBindingsMap[clusterStatus.ClusterName]
+		if !condition.IsConditionStatusTrue(clusterStartedCond, updateRun.Generation) {
+			// The cluster has not started updating yet.
+			if !isBindingSyncedWithClusterStatus(updateRun, binding, clusterStatus) {
+				klog.V(2).InfoS("Found the first cluster that needs to be updated", "cluster", clusterStatus.ClusterName, "stage", updatingStageStatus.StageName, "clusterStagedUpdateRun", updateRunRef)
+				// The binding is not up-to-date with the cluster status.
+				binding.Spec.State = placementv1beta1.BindingStateBound
+				binding.Spec.ResourceSnapshotName = resourceSnapshotName
+				binding.Spec.ResourceOverrideSnapshots = clusterStatus.ResourceOverrideSnapshots
+				binding.Spec.ClusterResourceOverrideSnapshots = clusterStatus.ClusterResourceOverrideSnapshots
+				binding.Spec.ApplyStrategy = updateRun.Status.ApplyStrategy
+				if err := r.Client.Update(ctx, binding); err != nil {
+					klog.ErrorS(err, "Failed to update binding to be bound with the matching spec of the updateRun", "binding", klog.KObj(binding), "clusterStagedUpdateRun", updateRunRef)
+					return 0, controller.NewUpdateIgnoreConflictError(err)
+				}
+				klog.V(2).InfoS("Updated the status of a binding to bound", "binding", klog.KObj(binding), "cluster", clusterStatus.ClusterName, "stage", updatingStageStatus.StageName, "clusterStagedUpdateRun", updateRunRef)
+			} else {
+				klog.V(2).InfoS("Found the first binding that is updating but the cluster status has not been updated", "cluster", clusterStatus.ClusterName, "stage", updatingStageStatus.StageName, "clusterStagedUpdateRun", updateRunRef)
+				if binding.Spec.State != placementv1beta1.BindingStateBound {
+					binding.Spec.State = placementv1beta1.BindingStateBound
+					if err := r.Client.Update(ctx, binding); err != nil {
+						klog.ErrorS(err, "Failed to update a binding to be bound", "binding", klog.KObj(binding), "cluster", clusterStatus.ClusterName, "stage", updatingStageStatus.StageName, "clusterStagedUpdateRun", updateRunRef)
+						return 0, controller.NewUpdateIgnoreConflictError(err)
+					}
+					klog.V(2).InfoS("Updated the status of a binding to bound", "binding", klog.KObj(binding), "cluster", clusterStatus.ClusterName, "stage", updatingStageStatus.StageName, "clusterStagedUpdateRun", updateRunRef)
+				} else {
+					if _, updateErr := checkClusterUpdateResult(binding, clusterStatus, updatingStageStatus, updateRun); updateErr != nil {
```

Comment:
For now, yes. Because we currently do not identify any termination errors for cluster updates. I have a TODO in this function. But even in the future when we can identify some, I'll return an `errExecutionAborted` inside the `checkClusterUpdateResult` function and can still return the `updateErr` directly.